### PR TITLE
ASR: Sync Libasr

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
This PR syncs the current `Libasr` of `LFortran` and `LPython`.

Corresponding MR at `LFortran`: https://gitlab.com/lfortran/lfortran/-/merge_requests/1816

Corresponding PR at `Libasr`: https://github.com/lcompilers/libasr/pull/2

**Notes:**
- Reference tests for both `LFortran` and `LPython` updated.
    - For `LPython`, it seems there is change in reference tests of `cpp` backend only.
- At few places in the code (exactly `4` places, I think), there is a comment starting with `Sync:` indicating slight doubt in the respective `code`/`statement`
- For `LPython`: All tests pass.
- For `LFortran`: Few tests (5 failing tests in `tests/tests.toml` and 4 failing tests in `integration_tests/CMakeLists.txt`) are commented out. The comments are marked with `Sync:` (to ease lookup / to mark as failing due to sync).

### CODE REVIEW

File: src/libasr/asr_utils.h [hunk 1]
```diff
@@ -38,49 +38,6 @@ static inline ASR::ttype_t* TYPE(const ASR::asr_t *f)
...
-        case ASR::symbolType::AssociateBlock: {
-            return ASR::down_cast<ASR::AssociateBlock_t>(f)->m_name;
-        }
-        case ASR::symbolType::Block: {
-            return ASR::down_cast<ASR::Block_t>(f)->m_name;
-        }
-        default : throw LFortranException("Not implemented");
-    }
-}
-
```
Comment (czgdp1807) [inline src/libasr/asr_utils.h:83]: Why is this removed? This is needed.

File: src/libasr/asr_utils.h [hunk 2]
```diff
@@ -1022,6 +1032,21 @@ static inline ASR::ttype_t* duplicate_type(Allocator& al, const ASR::ttype_t* t,
...
+    }
+    if( is_source_pointer && !is_dest_pointer ) {
+        ASR::ttype_t* temp = source;
+        source = dest;
+        dest = temp;
+    }
+    bool res = source->type == ASR::down_cast<ASR::Pointer_t>(dest)->m_type->type;
+    return res;
+}
+
```
Comment (czgdp1807) [inline src/libasr/asr_utils.h:1049]: Not needed. Please remove it.

Comment (czgdp1807) [inline src/libasr/asr_utils.h:1049]:
```suggestion
```

File: src/libasr/codegen/asr_to_c.cpp [hunk 3]
```diff
@@ -94,30 +126,40 @@ class ASRToCVisitor : public BaseCCPPVisitor<ASRToCVisitor>
...
+                std::string type_name;
+                if (t->m_kind == 1) {
+                    type_name = "int8_t";
+                } else if (t->m_kind == 2) {
+                    type_name = "int16_t";
+                } else if (t->m_kind == 4) {
+                    type_name = "int32_t";
+                } else if (t->m_kind == 8) {
+                    type_name = "int64_t";
+                }
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_c.cpp:143]: The earlier code was better. I removed this and replaced it with the one which is deleted. Please restore.

File: src/libasr/codegen/asr_to_c.cpp [hunk 4]
```diff
@@ -94,30 +126,40 @@ class ASRToCVisitor : public BaseCCPPVisitor<ASRToCVisitor>
...
                 std::string type_name = "float complex";
                 if (t->m_kind == 8) type_name = "double complex";
                 sub = format_type_c(dims, type_name, v.m_name, use_ref, dummy);
             } else if (ASRUtils::is_logical(*v.m_type)) {
                 ASR::Logical_t *t = ASR::down_cast<ASR::Logical_t>(v.m_type);
-                std::string dims = convert_dims_c(t->n_dims, t->m_dims);
+                dims = convert_dims_c(t->n_dims, t->m_dims);
                 sub = format_type_c(dims, "bool", v.m_name, use_ref, dummy);
-            } else if (ASRUtils::is_character(*v.m_type)) {
+            }else if (ASRUtils::is_character(*v.m_type)) {
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_c.cpp:162]:
```suggestion
            } else if (ASRUtils::is_character(*v.m_type)) {
```

File: src/libasr/asr_utils.h (hunk 1 above)
Comment (Shaikh-Ubaid) [reply src/libasr/asr_utils.h:83]: I guess this comment is about the function `static inline char *symbol_name(const ASR::symbol_t *f)` (from [here](https://github.com/lcompilers/lpython/pull/730/files/c248ae1b6112b2a0436adab2327b5742de4ab586#diff-da3dce35379cdaaf286c90ae38fe3d27efda9a299f60c682dacc97ad1eca5ce7L41)).

It is defined below [here](https://github.com/lcompilers/lpython/pull/730/files/c248ae1b6112b2a0436adab2327b5742de4ab586#diff-da3dce35379cdaaf286c90ae38fe3d27efda9a299f60c682dacc97ad1eca5ce7R193). It is defined just above `static inline SymbolTable *symbol_parent_symtab(const ASR::symbol_t *f)` ([here](https://github.com/lcompilers/lpython/pull/730/files/c248ae1b6112b2a0436adab2327b5742de4ab586#diff-da3dce35379cdaaf286c90ae38fe3d27efda9a299f60c682dacc97ad1eca5ce7L226)) and this convention is taken from `LFortran`. (In `LFortran`, the function `static inline char *symbol_name(const ASR::symbol_t *f)` was defined just above `static inline SymbolTable *symbol_parent_symtab(const ASR::symbol_t *f)`, so the same convention is followed here).

Comment (czgdp1807) [reply src/libasr/asr_utils.h:83]: Yes. That's why I minimized my comment after noticing that later.

**Comment by Shaikh-Ubaid** [comment]
@czgdp1807 I am working on the suggested changes. 

Please, could you possibly share why few test cases are not working for `LFortran` and what might be a possible fix for it, please?

**Comment by czgdp1807** [comment]
I will push the changes directly to your branch in both LFortran and LPython.

File: src/libasr/pass/pass_manager.h [hunk 5]
```diff
@@ -31,6 +39,8 @@
 #include <map>
 #include <vector>
 
+#include <iostream>
+
```
Comment (czgdp1807) [inline src/libasr/pass/pass_manager.h:43]: Remove this not needed.

Comment (Shaikh-Ubaid) [reply src/libasr/pass/pass_manager.h]: I guess this is about removing the `#include <iostream>` header file from `src/libasr/pass/pass_manager.h`. I just added for the same.

**Comment by Shaikh-Ubaid** [comment]
@czgdp1807 We removed `is_same_type_pointer()` during commit [05836a4](https://github.com/lcompilers/lpython/pull/730/commits/05836a4a1b3defe9929df7c5f8eebc7e68606789).

But at `LFortran`, it seems it is being needed:
```bash
In file included from /home/ubaid/OpenSource/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp:15:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/asr_implicit_cast_rules.h: In static member function ‘static void LFortran::ImplicitCastRules::set_converted_value(Allocator&, const LFortran::Location&, LFortran::ASR::expr_t**, LFortran::ASR::ttype_t*, LFortran::ASR::ttype_t*)’:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/asr_implicit_cast_rules.h:99:19: error: ‘is_same_type_pointer’ is not a member of ‘LFortran::ASRUtils’
   99 |         ASRUtils::is_same_type_pointer(source_type, dest_type)) {
      |                   ^~~~~~~~~~~~~~~~~~~~
In file included from /home/ubaid/OpenSource/lfortran/src/lfortran/semantics/ast_to_asr.cpp:12:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/asr_implicit_cast_rules.h: In static member function ‘static void LFortran::ImplicitCastRules::set_converted_value(Allocator&, const LFortran::Location&, LFortran::ASR::expr_t**, LFortran::ASR::ttype_t*, LFortran::ASR::ttype_t*)’:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/asr_implicit_cast_rules.h:99:19: error: ‘is_same_type_pointer’ is not a member of ‘LFortran::ASRUtils’
   99 |         ASRUtils::is_same_type_pointer(source_type, dest_type)) {
      |                   ^~~~~~~~~~~~~~~~~~~~
In file included from /home/ubaid/OpenSource/lfortran/src/lfortran/semantics/ast_body_visitor.cpp:14:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/asr_implicit_cast_rules.h: In static member function ‘static void LFortran::ImplicitCastRules::set_converted_value(Allocator&, const LFortran::Location&, LFortran::ASR::expr_t**, LFortran::ASR::ttype_t*, LFortran::ASR::ttype_t*)’:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/asr_implicit_cast_rules.h:99:19: error: ‘is_same_type_pointer’ is not a member of ‘LFortran::ASRUtils’
   99 |         ASRUtils::is_same_type_pointer(source_type, dest_type)) {
      |                   ^~~~~~~~~~~~~~~~~~~~
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/ast_body_visitor.cpp: In member function ‘void LFortran::BodyVisitor::visit_Associate(const LFortran::AST::Associate_t&)’:
/home/ubaid/OpenSource/lfortran/src/lfortran/semantics/ast_body_visitor.cpp:480:23: error: ‘is_same_type_pointer’ is not a member of ‘LFortran::ASRUtils’
  480 |         if( ASRUtils::is_same_type_pointer(target_type, value_type) ) {
      |                       ^~~~~~~~~~~~~~~~~~~~
```

**Comment by Shaikh-Ubaid** [comment]
Since, `LFortran` is needing `is_same_type_pointer()`, shall we add `is_same_type_pointer()` back for this `PR`? Maybe we could remove it in future `PRs`

**Comment by czgdp1807** [comment]
Sure, makes sense.

**Comment by Shaikh-Ubaid** [comment]
Currently all the three repositories use the same `libasr`.

Corresponding comment at LFortran: https://gitlab.com/lfortran/lfortran/-/merge_requests/1816

Corresponding comment at Libasr: https://github.com/lcompilers/libasr/pull/2
